src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingEndpoint.cs

```diff
@@ -147,6 +184,120 @@ public async Task<TextEditContainer> Handle(DocumentRangeFormattingParams reques
+                return null;
+            }
+
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
+            var hostDocumentIndex = request.Position.GetAbsoluteIndex(sourceText);
+            var triggerCharacterKind = _razorDocumentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex);
+            if (triggerCharacterKind is not RazorLanguageKind.CSharp)
+            {
+                _logger.LogInformation($"Unsupported trigger character language {triggerCharacterKind:G}.");
+                return null;
+            }
+
+            if (!IsApplicableTriggerCharacter(request.Character, triggerCharacterKind))
+            {
+                // We were triggered but the trigger character doesn't make sense for the current cursor position. Bail.
+                _logger.LogInformation("Unsupported trigger character location.");
+                return null;
+            }
+
+            if (!_razorDocumentMappingService.TryMapToProjectedDocumentPosition(codeDocument, hostDocumentIndex, out _, out var projectedIndex))
+            {
+                _logger.LogWarning($"Failed to map to projected position for document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var formattingContext = FormattingContext.Create(
+                request.TextDocument.Uri, documentSnapshot, codeDocument, request.Options, _adhocWorkspaceFactory, isFormatOnType: true);
+            var documentOptions = await GetDocumentOptionsAsync(request, formattingContext.CSharpWorkspaceDocument).ConfigureAwait(false);
+
+            // Ask C# for formatting changes.
+            var formattingChanges = await RazorCSharpFormattingInteractionService.GetFormattingChangesAsync(
+                formattingContext.CSharpWorkspaceDocument, typedChar: request.Character[0], projectedIndex, documentOptions, cancellationToken).ConfigureAwait(false);
+
+            formattingContext.Dispose();
+
+            if (formattingChanges.IsEmpty)
+            {
+                _logger.LogInformation("Received no results.");
+                return null;
+            }
+
+            var textEdits = formattingChanges.Select(
+                change => Extensions.TextChangeExtensions.AsTextEdit(change, codeDocument.GetCSharpSourceText())).ToArray();
+            _logger.LogInformation($"Received {textEdits.Length} results from C#.");
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var formattedEdits = await _razorFormattingService.ApplyFormattedEditsAsync(
+                request.TextDocument.Uri, documentSnapshot, triggerCharacterKind, textEdits, request.Options, cancellationToken).ConfigureAwait(false);
+
+            _logger.LogInformation($"Returning {formattedEdits.Length} final formatted results.");
+            return formattedEdits;
+        }
 #nullable restore
```

Comment:
Needed?

Reply:
I believe it's needed since we disable nullable [earlier in the file](https://github.com/dotnet/aspnetcore-tooling/blob/bc85a5abf038256bfbc4ff54578947ed9c8d6696/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingEndpoint.cs#L156) due to O# constraints.